### PR TITLE
feat: customize nav bar background color

### LINKED ISSUE #1451: UI: Add ability to visually distinguish between deployment environments
**Is this a BUG REPORT or FEATURE REQUEST?**: **FEATURE REQUEST**

We run a dev environment and a production environment, which is fairly typical. We currently access the `argo-ui` through a proxy cmd, and I think it's easy enough to not specific the environment (dev/prod) properly and get confused as both look visually the same. 

What do you think about this problem? If/When we place domains in front of the UIs,  eg. `dev-argo-ui.company-ds.com` and `prod-argo-ui.company-ds.com` I think this issue will be alleviated, but not totally fixed as no element exists in the UI to display the deployment environment.

### PR DESCRIPTION
**Pull request by krrrr38**
Signed-off-by: krrrr38 <k.kaizu38@gmail.com>

* [x] Run `make pre-commit -B` to fix codegen, lint, and commit message problems.

## Motivation 

I want to distinguish multiple argo-workflows easily not to operate invalid operation on different environments.

Fixes #1451

## How to solve

- Make argo-ui NavBar customizable to change nav bar style
  - https://github.com/argoproj/argo-ui/pull/167
- Add navigation bar background color configuration into argo-workflow config map
- Pass background color info from `GetInfo` api

## Discussions

There are some discussions about this feature.
- [argo-cd way customize css impl on argo-workflows](https://github.com/argoproj/argo-workflows/pull/6282)
- argo-workflows should provide argo-workflow way.
  - [customize css sometimes breaks the ui, so provide argo-workflow `theme` and css maintained by argo](https://github.com/argoproj/argo-workflows/pull/6282#pullrequestreview-739570927)
- [More pluggable styles](https://github.com/argoproj/argo-workflows/issues/7054)

To archive my goal, I add nav bar color config for following reasons.

- [argo-ui v2 has `theme` feature which support only `dark` or `style`](https://github.com/argoproj/argo-ui/blob/49f257794525a5d6051bff109147b6133c7ee550/v2/shared/context/theme.tsx)
  - this theme feature doesn't enough to distinguish multiple environments
- When using `theme`, to distinguish multiple environments, we need to provides 3?, 4?,... themes. (dev, stg, prd, e2e, x division? x teams?,...)
  - I have no skills to create new design argo themes..., but color settings are available too many colors and easy to distinguish.
- pluggable ui and nav bar background color are something different.
  - pluggable ui provides new feature which provides **new component** in specific domains.
    - (if this is not, sorry 🙇 )
  - nav bar background color is **global settings** which is useful to distinguish multiple environments
- Not to customize too many css, just add a color setting.

So, I add nav bar background color as **global argo-workflow ui config**. I hope this `ui` config might be support icon, title, and so on.

## Check

https://github.com/argoproj/argo-workflows/pull/7387#issuecomment-1018985618

### CODE REVIEW

**Comment by krrrr38** [comment]
CI / UI failed because this feature requires https://github.com/argoproj/argo-ui/pull/167.

File: pkg/apis/workflow/v1alpha1/info.go [hunk 1]
```diff
@@ -11,3 +11,9 @@ type Link struct {
 	// The URL. Can contain "${metadata.namespace}", "${metadata.name}", "${status.startedAt}", "${status.finishedAt}" or any other element in workflow yaml, e.g. "${workflow.metadata.annotations.userDefinedKey}"
 	URL string `json:"url" protobuf:"bytes,3,opt,name=url"`
 }
+
+// Ui defines a config for UI
+type Ui struct {
+	// NavColor is an ui navigation bar theme which switch the design
+	NavColor string `json:"navColor,omitempty" protobuf:"bytes,1,opt,name=navColor"`
+}
```
Comment (simster7) [inline pkg/apis/workflow/v1alpha1/info.go:19]: If anything, wouldn't it be more powerful to allow a way to inject CSS overrides?

**Review by alexec — COMMENTED** [review]
The use case for this are pretty simple, the ability to set colour differently for different environments. I think this is probably good enough for that. Previous solutions have been, these have been complex, I like the simplicity

File: pkg/apis/workflow/v1alpha1/info.go [hunk 2]
```diff
@@ -11,3 +11,9 @@ type Link struct {
 	// The URL. Can contain "${metadata.namespace}", "${metadata.name}", "${status.startedAt}", "${status.finishedAt}" or any other element in workflow yaml, e.g. "${workflow.metadata.annotations.userDefinedKey}"
 	URL string `json:"url" protobuf:"bytes,3,opt,name=url"`
 }
+
+// Ui defines a config for UI
+type Ui struct {
```
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/info.go:16]: should be `UI` ,i.e. upper-case

**Review by alexec — CHANGES_REQUESTED** [review]

**Review by alexec — APPROVED** [review]

**Review by alexec — CHANGES_REQUESTED** [review]
I don't think we actually need the "UI" wrapper, as we already have `links` top-level. Can just be `info.navColor`

**Comment by krrrr38** [comment]
> I don't think we actually need the "UI" wrapper, as we already have `links` top-level. Can just be `info.navColor`

I also worry about this. I'll remove ui wrapper

File: pkg/apis/workflow/v1alpha1/info.go (hunk 1 above)
Comment (krrrr38) [reply pkg/apis/workflow/v1alpha1/info.go]: Based on the description`Discussions`, I think this way is better.

**Comment by krrrr38** [comment]
re-confirmed that nav bar color was changed by following config map

```
> kubectl get cm -n argo workflow-controller-configmap -ojsonpath='{.data.navColor}'
red
```

**Review by alexec — APPROVED** [review]

**Review by alexec — CHANGES_REQUESTED** [review]
Please fix build.

**Comment by krrrr38** [comment]
@alexec thank you for your review.
To build this PR, it requires argo-ui changes.
Can anyone review https://github.com/argoproj/argo-ui/pull/167?

**Comment by alexec** [comment]
Up-stream PR now merged, you should. be able to use argo-ui version v2.5.0.

**Comment by krrrr38** [comment]
I tried `yarn upgrade argo-ui@"https://github.com/argoproj/argo-ui.git"`, but there are some errors which depends on other library updates.
I will create another PR to upgrade argo-ui.

**Comment by krrrr38** [comment]
now works fine again.

image

File: server/info/info_server_test.go [hunk 3]
```diff
@@ -24,3 +25,31 @@ func Test_infoServer_GetUserInfo(t *testing.T) {
 		assert.Equal(t, "my-sa", info.ServiceAccountName)
 	}
 }
+
+func Test_infoServer_GetInfo(t *testing.T) {
```
Comment (alexec) [inline server/info/info_server_test.go:29]: nice

**Review by alexec — APPROVED** [review]
LGTM

**Comment by krrrr38** [comment]
@alexec sorry, fixed golang-ci 😇 . CI succeeded.

**Pull request merged**